compiler/rustc_typeck/src/coherence/inherent_impls_overlap.rs

```diff
@@ -113,9 +120,35 @@ impl ItemLikeVisitor<'v> for InherentOverlapChecker<'tcx> {
                 let ty_def_id = self.tcx.hir().local_def_id(item.hir_id);
                 let impls = self.tcx.inherent_impls(ty_def_id);
 
-                for (i, &impl1_def_id) in impls.iter().enumerate() {
-                    for &impl2_def_id in &impls[(i + 1)..] {
-                        if self.impls_have_common_items(impl1_def_id, impl2_def_id) {
+                // Perform a O(n^2) algorithm for small n,
+                // otherwise switch to an allocating algorithm with
+                // faster asymptotic runtime.
+                if impls.len() < 20 {
+                    for (i, &impl1_def_id) in impls.iter().enumerate() {
+                        for &impl2_def_id in &impls[(i + 1)..] {
```

Comment:
I guess this is how it is and there is no reason to change, but wouldn't cloning the iterator for the inner loop be nicer? There's no need to complicate the iteration when we only need to look ahead, I think.
Am I missing something?

Reply:
IDK for me as someone coming from C/C++, iterator magic is more complicated than doing things with indices.

Reply:
I also have a background in C and I hate working with indices when I can avoid them, they are too much a footgun :)